File: src/services/styles/styleProfileService.ts
/**
 * AI Novel Studio - 风格方案服务
 * Tauri 桌面端使用 SQLite，浏览器开发态使用 localStorage 回退。
 */
import { dbCall, getDbMode, lsGet, lsSet, generateId, nowISO } from '../database/db';
import type {
  StyleProfile,
  CreateStyleProfileInput,
  UpdateStyleProfileInput,
} from '../../types/style';

const STYLE_KEY = 'ai_novel_studio_style_profiles';

interface StyleProfileDto {
  id: string;
  projectId?: string | null;
  name: string;
  description?: string;
  narrativePerspective?: string;
  tone?: string;
  pace?: string;
  sentenceStyle?: string;
  dialogueRatio: number;
  descriptionRatio: number;
  psychologicalRatio?: number;
  battleStyle?: string;
  battleIntensity?: string;
  emotionTendency?: string;
  chapterEnding?: string;
  forbiddenStylesJson?: string;
  styleSummary?: string;
  rawConfigJson?: string;
  isActive: boolean;
  sourceType: string;
  sourceAssetId?: string;
  sourceReferenceWorkId?: string;
  sourceReferenceImportId?: string;
  sourceContentHash?: string;
  sourceState?: StyleProfile['sourceState'];
  analysisMetadataJson?: string;
  createdAt: string;
  updatedAt: string;
}

function fromDto(dto: StyleProfileDto): StyleProfile {
  let forbidden: string[] = [];
  try {
    forbidden = JSON.parse(dto.forbiddenStylesJson || '[]');
  } catch {
    /* ignore */
  }
  return {
    id: dto.id,
    novelId: dto.projectId?.trim() || undefined,
    name: dto.name,
    sourceType: dto.sourceType as StyleProfile['sourceType'],
    sourceAssetId: dto.sourceAssetId,
    sourceReferenceWorkId: dto.sourceReferenceWorkId,
    sourceReferenceImportId: dto.sourceReferenceImportId,
    sourceContentHash: dto.sourceContentHash,
    sourceState: dto.sourceState,
    analysisMetadataJson: dto.analysisMetadataJson,
    description: dto.description,
    narrativePerspective: dto.narrativePerspective,
    tone: dto.tone,
    pace: dto.pace,
    sentenceStyle: dto.sentenceStyle,
    dialogueRatio: dto.dialogueRatio,
    descriptionRatio: dto.descriptionRatio,
    psychologicalRatio: dto.psychologicalRatio,
    battleStyle: dto.battleStyle,
    battleIntensity: dto.battleIntensity,
    emotionTendency: dto.emotionTendency,
    chapterEnding: dto.chapterEnding,
    prohibitedStyles: forbidden,
    forbiddenStyles: forbidden,
    styleSummary: dto.styleSummary,
    rawConfigJson: dto.rawConfigJson,
    isActive: dto.isActive,
    targetWordsPerChapter: 4000,
    rhythmPreference: dto.pace === '快' ? 'fast' : dto.pace === '慢' ? 'slow' : 'moderate',
    createdAt: dto.createdAt,
    updatedAt: dto.updatedAt,
  };
}

const defaultSeed: CreateStyleProfileInput[] = [
  {
    name: '默认小说风格',
    sourceType: 'system_default',
    narrativePerspective: '第三人称有限视角',
    tone: '中性偏沉稳',
    pace: '中等',
    dialogueRatio: 0.35,
    descriptionRatio: 0.4,
    styleSummary: '适合大多数小说的通用风格配置。',
  },
  {
    name: '快节奏战斗风',
    sourceType: 'system_default',
    narrativePerspective: '第三人称',
    tone: '紧张、压迫',
    pace: '快',
    dialogueRatio: 0.25,
    descriptionRatio: 0.45,
    battleStyle: '重视动作连贯和代价',
    battleIntensity: 'high',
    emotionTendency: '紧张、压迫、爆发',
    styleSummary: '适合战斗密集的章节。',
  },
  {
    name: '抒情日常风',
    sourceType: 'system_default',
    narrativePerspective: '第三人称有限视角',
    tone: '温暖、细腻',
    pace: '慢',
    dialogueRatio: 0.4,
    descriptionRatio: 0.35,
    psychologicalRatio: 0.25,
    emotionTendency: '柔和、深情',
    styleSummary: '适合日常过渡和角色情感发展章节。',
  },
];

function getLocalStoredAll(): StyleProfile[] {
  return lsGet<StyleProfile[]>(STYLE_KEY) ?? [];
}

function getLocalAll(): StyleProfile[] {
  let list = getLocalStoredAll();
  if (list.length === 0) {
    const now = nowISO();
    list = defaultSeed.map((s, index) => ({
      ...s,
      id: generateId(),
      sourceType: 'system_default' as StyleProfile['sourceType'],
      targetWordsPerChapter: 4000,
      rhythmPreference:
        s.pace === '快'
          ? ('fast' as const)
          : s.pace === '慢'
            ? ('slow' as const)
            : ('moderate' as const),
      dialogueRatio: s.dialogueRatio || 0.35,
      descriptionRatio: s.descriptionRatio || 0.4,
      prohibitedStyles: [],
      isActive: index === 0,
      createdAt: now,
      updatedAt: now,
    }));
    lsSet(STYLE_KEY, list);
  }
  return list;
}

export const styleProfileService = {
  async getAll(novelId?: string, options: { initialize?: boolean } = {}): Promise<StyleProfile[]> {
    const initialize = options.initialize !== false;
    const readLocal = initialize ? getLocalAll : getLocalStoredAll;
    try {
      const dtos = await dbCall<StyleProfileDto[]>(
        'list_style_profiles',
        { projectId: novelId },
        () => {
          return readLocal().map((s): StyleProfileDto => ({
            id: s.id,
            projectId: s.novelId || '',
            name: s.name,
            description: s.description,
            narrativePerspective: s.narrativePerspective,
            tone: s.tone,
            pace: s.pace,
            sentenceStyle: s.sentenceStyle,
            dialogueRatio: s.dialogueRatio,
            descriptionRatio: s.descriptionRatio,
            psychologicalRatio: s.psychologicalRatio,
            battleStyle: s.battleStyle,
            battleIntensity: s.battleIntensity,
            emotionTendency: s.emotionTendency,
            chapterEnding: s.chapterEnding,
            forbiddenStylesJson: JSON.stringify(s.prohibitedStyles || []),
            styleSummary: s.styleSummary,
            rawConfigJson: s.rawConfigJson,
            isActive: s.isActive,
            sourceType: s.sourceType,
            sourceAssetId: s.sourceAssetId,
            sourceReferenceWorkId: s.sourceReferenceWorkId,
            sourceReferenceImportId: s.sourceReferenceImportId,
            sourceContentHash: s.sourceContentHash,
            sourceState: s.sourceState,
            analysisMetadataJson: s.analysisMetadataJson,
            createdAt: s.createdAt,
            updatedAt: s.updatedAt,
          }));
        },
      );
      const profiles = dtos.map(fromDto);
      if (profiles.length > 0)
        return novelId ? profiles.filter((s) => !s.novelId || s.novelId === novelId) : profiles;
      if (getDbMode() === 'tauri') return [];
      return novelId
        ? readLocal().filter((profile) => !profile.novelId || profile.novelId === novelId)
        : readLocal();
    } catch (error) {
      if (getDbMode() === 'tauri') throw error;
      return novelId ? readLocal().filter((s) => !s.novelId || s.novelId === novelId) : readLocal();
    }
  },

  async getActive(novelId: string): Promise<StyleProfile | null> {
    try {
      const dto = await dbCall<StyleProfileDto | null>(
        'get_active_style_profile',
        { projectId: novelId },
        () => null,
      );
      return dto ? fromDto(dto) : null;
    } catch (error) {
      if (getDbMode() === 'tauri') throw error;
      const local = getLocalAll();
      return (
        local.find((profile) => profile.novelId === novelId && profile.isActive) ??
        local.find((profile) => !profile.novelId && profile.isActive) ??
        local.find((profile) => profile.novelId === novelId) ??
        local.find((profile) => !profile.novelId) ??
        null
      );
    }
  },

  async getById(id: string): Promise<StyleProfile | null> {
    try {
      const all = await this.getAll();
      return all.find((s) => s.id === id) ?? null;
    } catch (error) {
      if (getDbMode() === 'tauri') throw error;
      return getLocalAll().find((s) => s.id === id) ?? null;
    }
  },

  async create(
    input: CreateStyleProfileInput & { projectId?: string; novelId?: string },
  ): Promise<StyleProfile> {
    const pid = input.projectId || input.novelId || '';
    try {
      const dto = await dbCall<StyleProfileDto>(
        'save_style_profile',
        {
          id: undefined,
          input: {
            projectId: pid,
            name: input.name,
            narrativePerspective: input.narrativePerspective,
            tone: input.tone,
            pace: input.pace,
            sentenceStyle: input.sentenceStyle,
            dialogueRatio: input.dialogueRatio,
            descriptionRatio: input.descriptionRatio,
            psychologicalRatio: input.psychologicalRatio,
            battleStyle: input.battleStyle,
            battleIntensity: input.battleIntensity,
            emotionTendency: input.emotionTendency,
            chapterEnding: input.chapterEnding,
            forbiddenStyles: input.forbiddenStyles || [],
            styleSummary: input.styleSummary,
            sourceType: input.sourceType || 'manual',
            sourceAssetId: input.sourceAssetId,
            sourceReferenceWorkId: input.sourceReferenceWorkId,
            sourceReferenceImportId: input.sourceReferenceImportId,
            sourceContentHash: input.sourceContentHash,
            sourceState: input.sourceState,
            analysisMetadataJson: input.analysisMetadataJson,
          },
        },
        () => {
          throw new Error('fallback');
        },
      );
      return fromDto(dto);
    } catch (error) {
      if (getDbMode() === 'tauri') throw error;
      const list = getLocalAll();
      const now = nowISO();
      const profile: StyleProfile = {
        ...input,
        id: generateId(),
        novelId: pid,
        sourceType: input.sourceType || 'manual',
        targetWordsPerChapter: 4000,
        rhythmPreference: input.pace === '快' ? 'fast' : input.pace === '慢' ? 'slow' : 'moderate',
        dialogueRatio: input.dialogueRatio || 0.35,
        descriptionRatio: input.descriptionRatio || 0.4,
        prohibitedStyles: input.forbiddenStyles || [],
        isActive: false,
        createdAt: now,
        updatedAt: now,
      };
      list.push(profile);
      lsSet(STYLE_KEY, list);
      return profile;
    }
  },

  async update(
    id: string,
    input: UpdateStyleProfileInput & { projectId?: string; novelId?: string },
  ): Promise<StyleProfile | null> {
    const existing = await this.getById(id);
    if (!existing) return null;
    const requestedOwner = input.projectId ?? input.novelId;
    const pid = requestedOwner?.trim() ?? existing.novelId ?? '';
    try {
      const dto = await dbCall<StyleProfileDto>(
        'save_style_profile',
        {
          id,
          input: {
            projectId: pid,
            name: input.name ?? existing.name,
            description: existing.description,
            narrativePerspective: input.narrativePerspective ?? existing.narrativePerspective,
            tone: input.tone ?? existing.tone,
            pace: input.pace ?? existing.pace,
            sentenceStyle: input.sentenceStyle ?? existing.sentenceStyle,
            dialogueRatio: input.dialogueRatio ?? existing.dialogueRatio,
            descriptionRatio: input.descriptionRatio ?? existing.descriptionRatio,
            psychologicalRatio: input.psychologicalRatio ?? existing.psychologicalRatio,
            battleStyle: input.battleStyle ?? existing.battleStyle,
            battleIntensity: input.battleIntensity ?? existing.battleIntensity,
            emotionTendency: input.emotionTendency ?? existing.emotionTendency,
            chapterEnding: input.chapterEnding ?? existing.chapterEnding,
            forbiddenStyles:
              input.forbiddenStyles ?? existing.forbiddenStyles ?? existing.prohibitedStyles,
            styleSummary: input.styleSummary ?? existing.styleSummary,
            rawConfigJson: existing.rawConfigJson,
            sourceType: input.sourceType ?? existing.sourceType,
            sourceAssetId: input.sourceAssetId ?? existing.sourceAssetId,
            sourceReferenceWorkId: input.sourceReferenceWorkId ?? existing.sourceReferenceWorkId,
            sourceReferenceImportId:
              input.sourceReferenceImportId ?? existing.sourceReferenceImportId,
            sourceContentHash: input.sourceContentHash ?? existing.sourceContentHash,
            sourceState: input.sourceState ?? existing.sourceState,
            analysisMetadataJson: input.analysisMetadataJson ?? existing.analysisMetadataJson,
          },
        },
        () => {
          throw new Error('fallback');
        },
      );
      return fromDto(dto);
    } catch (error) {
      if (getDbMode() === 'tauri') throw error;
      const list = getLocalAll();
      const idx = list.findIndex((s) => s.id === id);
      if (idx === -1) return null;
      list[idx] = { ...list[idx], ...input, updatedAt: nowISO() };
      if (input.forbiddenStyles) list[idx].prohibitedStyles = input.forbiddenStyles;
      lsSet(STYLE_KEY, list);
      return list[idx];
    }
  },

  async setActive(projectId: string, styleProfileId: string): Promise<void> {
    const normalizedProjectId = projectId.trim();
    await dbCall<void>(
      'set_active_style_profile',
      { input: { projectId: normalizedProjectId, styleProfileId } },
      () => {
        const list = getLocalAll();
        const target = list.find((profile) => profile.id === styleProfileId);
        if (!target || (target.novelId ?? '').trim() !== normalizedProjectId) {
          throw new Error('风格方案不存在或不属于当前作品');
        }
        for (const profile of list) {
          if ((profile.novelId ?? '').trim() === normalizedProjectId) {
            profile.isActive = profile.id === styleProfileId;
          }
        }
        lsSet(STYLE_KEY, list);
      },
    );
  },

  async remove(projectIdOrId: string, id?: string): Promise<void> {
    const sid = id || projectIdOrId;
    const existing = id ? null : await this.getById(sid);
    if (!id && !existing) return;
    const pid = id ? projectIdOrId : (existing?.novelId ?? '');
    await dbCall<void>('delete_style_profile', { projectId: pid, styleProfileId: sid }, () => {
      lsSet(
        STYLE_KEY,
        getLocalAll().filter((s) => s.id !== sid),
      );
    });
  },
};

File: src-tauri/src/services/style_profile_service.rs
use crate::domain::ai::{SaveStyleProfileInput, SetActiveStyleProfileInput, StyleProfileDto};
use crate::repositories::style_profile_repository;
use rusqlite::{params, Connection, OptionalExtension};

pub fn list_style_profiles(
    conn: &Connection,
    project_id: Option<&str>,
) -> Result<Vec<StyleProfileDto>, String> {
    style_profile_repository::list_style_profiles(conn, project_id)
}

pub fn get_active_style_profile(
    conn: &Connection,
    project_id: &str,
) -> Result<Option<StyleProfileDto>, String> {
    style_profile_repository::find_active_style_profile(conn, project_id)
}

pub fn save_style_profile(
    conn: &Connection,
    id: Option<String>,
    input: SaveStyleProfileInput,
) -> Result<StyleProfileDto, String> {
    let now = chrono::Utc::now().to_rfc3339();
    let project_id = input.project_id.trim().to_string();
    let project_id = if project_id.is_empty() {
        None
    } else {
        Some(project_id)
    };
    let forbidden_json = serde_json::to_string(&input.forbidden_styles.unwrap_or_default())
        .unwrap_or_else(|_| "[]".to_string());
    let source_type = input.source_type.unwrap_or_else(|| "manual".to_string());
    let source_state = input.source_state.clone().unwrap_or_else(|| {
        if input.source_reference_import_id.is_some() {
            "available".to_string()
        } else {
            "none".to_string()
        }
    });
    if !matches!(
        source_state.as_str(),
        "none" | "available" | "outdated" | "missing" | "legacy_unverified"
    ) {
        return Err("REFERENCE_INPUT_INVALID: invalid style source state".to_string());
    }
    if let Some(metadata) = input.analysis_metadata_json.as_deref() {
        let parsed = serde_json::from_str::<serde_json::Value>(metadata)
            .map_err(|_| "REFERENCE_INPUT_INVALID: invalid analysis metadata".to_string())?;
        if !parsed.is_object() || metadata.len() > 500_000 {
            return Err("REFERENCE_INPUT_INVALID: invalid analysis metadata".to_string());
        }
    }
    let reference_fields = [
        input.source_reference_work_id.as_deref(),
        input.source_reference_import_id.as_deref(),
        input.source_content_sha256.as_deref(),
    ];
    let populated_reference_fields = reference_fields
        .iter()
        .filter(|value| value.is_some())
        .count();
    if populated_reference_fields != 0 && populated_reference_fields != reference_fields.len() {
        return Err("REFERENCE_INPUT_INVALID: incomplete style reference identity".to_string());
    }
    if populated_reference_fields == reference_fields.len() {
        let source_hash = input.source_content_sha256.as_deref().unwrap_or_default();
        if source_hash.len() != 64
            || !source_hash
                .bytes()
                .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
        {
            return Err("REFERENCE_INPUT_INVALID: invalid style source hash".to_string());
        }
        let valid_scope: i64 = conn
            .query_row(
                "SELECT COUNT(*) FROM reference_imports i
                 INNER JOIN reference_works w
                   ON w.id = i.reference_work_id AND w.novel_id = i.novel_id
                 WHERE w.novel_id = ?1 AND w.id = ?2 AND i.id = ?3 AND i.source_sha256 = ?4
                   AND (?5 <> 'available' OR i.is_current = 1)",
                params![
                    project_id.as_deref().unwrap_or_default(),
                    input.source_reference_work_id,
                    input.source_reference_import_id,
                    input.source_content_sha256,
                    source_state,
                ],
                |row| row.get(0),
            )
            .map_err(|error| error.to_string())?;
        if valid_scope != 1 {
            return Err(
                "REFERENCE_SCOPE_MISMATCH: style reference does not belong to project".to_string(),
            );
        }
    }
    let source_state_update = if populated_reference_fields == reference_fields.len() {
        Some(source_state.clone())
    } else {
        input.source_state.clone()
    };

    if let Some(existing_id) = id {
        let updated = conn.execute(
            "UPDATE style_profiles SET name = ?1, description = ?2, narrative_perspective = ?3, tone = ?4, pace = ?5, sentence_style = ?6, dialogue_ratio = ?7, description_ratio = ?8, psychological_ratio = ?9, battle_style = ?10, battle_intensity = ?11, emotion_tendency = ?12, chapter_ending = ?13, forbidden_styles = ?14, style_summary = ?15, raw_config_json = ?16, source_type = ?17, updated_at = ?18, source_asset_id = COALESCE(?21, source_asset_id), source_reference_work_id = COALESCE(?22, source_reference_work_id), source_reference_import_id = COALESCE(?23, source_reference_import_id), source_content_sha256 = COALESCE(?24, source_content_sha256), source_state = COALESCE(?25, source_state), analysis_metadata_json = COALESCE(?26, analysis_metadata_json) WHERE id = ?19 AND ((?20 IS NULL AND novel_id IS NULL) OR novel_id = ?20)",
            params![
                &input.name, &input.description,
                &input.narrative_perspective, &input.tone, &input.pace, &input.sentence_style,
                input.dialogue_ratio.unwrap_or(0.35), input.description_ratio.unwrap_or(0.4),
                input.psychological_ratio, &input.battle_style, &input.battle_intensity,
                &input.emotion_tendency, &input.chapter_ending,
                &forbidden_json, &input.style_summary, &input.raw_config_json,
                &source_type, &now, &existing_id, &project_id,
                &input.source_asset_id, &input.source_reference_work_id,
                &input.source_reference_import_id, &input.source_content_sha256,
                &source_state_update, &input.analysis_metadata_json,
            ],
        ).map_err(|e| e.to_string())?;
        if updated != 1 {
            return Err("风格方案不存在或不属于当前作品".to_string());
        }
        style_profile_repository::find_style_profile_by_id(conn, &existing_id)
    } else {
        let new_id = uuid::Uuid::new_v4().to_string();
        conn.execute(
            "INSERT INTO style_profiles (id, novel_id, name, description, narrative_perspective, tone, pace, sentence_style, dialogue_ratio, description_ratio, psychological_ratio, battle_style, battle_intensity, emotion_tendency, chapter_ending, forbidden_styles, style_summary, is_active, raw_config_json, source_type, created_at, updated_at, source_asset_id, source_reference_work_id, source_reference_import_id, source_content_sha256, source_state, analysis_metadata_json) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12,?13,?14,?15,?16,?17,0,?18,?19,?20,?20,?21,?22,?23,?24,?25,?26)",
            params![
                &new_id, &project_id, &input.name, &input.description,
                &input.narrative_perspective, &input.tone, &input.pace, &input.sentence_style,
                input.dialogue_ratio.unwrap_or(0.35), input.description_ratio.unwrap_or(0.4),
                input.psychological_ratio, &input.battle_style, &input.battle_intensity,
                &input.emotion_tendency, &input.chapter_ending,
                &forbidden_json, &input.style_summary, &input.raw_config_json,
                &source_type, &now,
                &input.source_asset_id, &input.source_reference_work_id,
                &input.source_reference_import_id, &input.source_content_sha256,
                &source_state, &input.analysis_metadata_json,
            ],
        ).map_err(|e| e.to_string())?;
        style_profile_repository::find_style_profile_by_id(conn, &new_id)
    }
}

pub fn set_active_style_profile(
    conn: &Connection,
    input: SetActiveStyleProfileInput,
) -> Result<(), String> {
    let expected_owner = if input.project_id.trim().is_empty() {
        None
    } else {
        Some(input.project_id.trim().to_string())
    };
    let owner = conn
        .query_row(
            "SELECT novel_id FROM style_profiles WHERE id = ?1",
            params![input.style_profile_id],
            |row| row.get::<_, Option<String>>(0),
        )
        .optional()
        .map_err(|error| error.to_string())?;
    if owner != Some(expected_owner) {
        return Err("风格方案不存在或不属于当前作品".to_string());
    }
    let now = chrono::Utc::now().to_rfc3339();
    style_profile_repository::select_active_style_profile_for_scope(
        conn,
        &input.style_profile_id,
        &input.project_id,
        &now,
    )
}

pub fn delete_style_profile(
    conn: &Connection,
    project_id: &str,
    style_profile_id: &str,
) -> Result<(), String> {
    let was_active =
        style_profile_repository::delete_style_profile_row(conn, style_profile_id, project_id)?;
    if was_active {
        if let Some(new_active_id) =
            style_profile_repository::find_latest_remaining_style_profile_id(conn, project_id)?
        {
            let now = chrono::Utc::now().to_rfc3339();
            let _ = style_profile_repository::activate_style_profile_row(
                conn,
                &new_active_id,
                project_id,
                &now,
            );
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use rusqlite::Connection;

    fn setup_test_db() -> Connection {
        let mut conn = Connection::open_in_memory().unwrap();
        crate::db::create_tables(&mut conn).unwrap();
        conn.execute(
            "INSERT INTO novels (id, title, created_at, updated_at) VALUES ('11111111-1111-1111-1111-111111111111', '测试小说', '2026-01-01T00:00:00Z', '2026-01-01T00:00:00Z')",
            [],
        ).unwrap();
        conn
    }

    #[test]
    fn test_style_profile_crud_and_activation() {
        let mut conn = setup_test_db();
        let input = SaveStyleProfileInput {
            project_id: "11111111-1111-1111-1111-111111111111".to_string(),
            name: "热血玄幻风".to_string(),
            description: Some("快节奏爽文风格".to_string()),
            narrative_perspective: Some("third".to_string()),
            tone: Some("passionate".to_string()),
            pace: Some("fast".to_string()),
            sentence_style: Some("concise".to_string()),
            dialogue_ratio: Some(0.4),
            description_ratio: Some(0.4),
            psychological_ratio: Some(0.2),
            battle_style: Some("detailed".to_string()),
            battle_intensity: Some("high".to_string()),
            emotion_tendency: Some("positive".to_string()),
            chapter_ending: Some("cliffhanger".to_string()),
            forbidden_styles: Some(vec!["水文".to_string()]),
            style_summary: Some("简洁明快，战斗刺激".to_string()),
            raw_config_json: None,
            source_type: Some("manual".to_string()),
            source_asset_id: None,
            source_reference_work_id: None,
            source_reference_import_id: None,
            source_content_sha256: None,
            source_state: None,
            analysis_metadata_json: None,
        };

        let profile = save_style_profile(&conn, None, input).unwrap();
        assert_eq!(profile.name, "热血玄幻风");
        assert_eq!(profile.is_active, false);

        set_active_style_profile(
            &conn,
            SetActiveStyleProfileInput {
                project_id: "11111111-1111-1111-1111-111111111111".to_string(),
                style_profile_id: profile.id.clone(),
            },
        )
        .unwrap();

        let active = get_active_style_profile(&conn, "11111111-1111-1111-1111-111111111111")
            .unwrap()
            .unwrap();
        assert_eq!(active.id, profile.id);
        crate::db::create_tables(&mut conn).unwrap();
        let active_after_seed =
            get_active_style_profile(&conn, "11111111-1111-1111-1111-111111111111")
                .unwrap()
                .unwrap();
        assert_eq!(active_after_seed.id, profile.id);

        delete_style_profile(&conn, "11111111-1111-1111-1111-111111111111", &profile.id).unwrap();
        let remaining =
            list_style_profiles(&conn, Some("11111111-1111-1111-1111-111111111111")).unwrap();
        assert_eq!(remaining.len(), 3);
        assert!(remaining.iter().all(|item| item.project_id.is_none()));
        let fallback = get_active_style_profile(&conn, "11111111-1111-1111-1111-111111111111")
            .unwrap()
            .unwrap();
        assert_eq!(fallback.name, "默认小说风格");
    }

    #[test]
    fn built_in_global_styles_are_seeded_idempotently() {
        let mut conn = Connection::open_in_memory().unwrap();
        conn.execute_batch("PRAGMA foreign_keys=ON;").unwrap();
        crate::db::create_tables(&mut conn).unwrap();
        crate::db::create_tables(&mut conn).unwrap();

        let profiles = list_style_profiles(&conn, None).unwrap();
        let built_ins = profiles
            .iter()
            .filter(|profile| profile.source_type == "system_default")
            .collect::<Vec<_>>();
        assert_eq!(built_ins.len(), 3);
        assert!(built_ins.iter().all(|profile| profile.project_id.is_none()));
        assert_eq!(
            built_ins
                .iter()
                .filter(|profile| profile.is_active)
                .map(|profile| profile.name.as_str())
                .collect::<Vec<_>>(),
            vec!["默认小说风格"]
        );
    }

    #[test]
    fn global_style_crud_uses_null_scope_and_rejects_cross_scope_activation() {
        let conn = setup_test_db();
        let input = SaveStyleProfileInput {
            project_id: String::new(),
            name: "全局冷峻风".to_string(),
            description: None,
            narrative_perspective: Some("第三人称有限视角".to_string()),
            tone: Some("冷峻".to_string()),
            pace: Some("中等".to_string()),
            sentence_style: None,
            dialogue_ratio: Some(0.3),
            description_ratio: Some(0.45),
            psychological_ratio: None,
            battle_style: None,
            battle_intensity: None,
            emotion_tendency: None,
            chapter_ending: None,
            forbidden_styles: Some(Vec::new()),
            style_summary: Some("克制、冷静。".to_string()),
            raw_config_json: None,
            source_type: Some("manual".to_string()),
            source_asset_id: None,
            source_reference_work_id: None,
            source_reference_import_id: None,
            source_content_sha256: None,
            source_state: None,
            analysis_metadata_json: None,
        };

        let profile = save_style_profile(&conn, None, input).unwrap();
        assert!(profile.project_id.is_none());
        set_active_style_profile(
            &conn,
            SetActiveStyleProfileInput {
                project_id: String::new(),
                style_profile_id: profile.id.clone(),
            },
        )
        .unwrap();
        let active = get_active_style_profile(&conn, "11111111-1111-1111-1111-111111111111")
            .unwrap()
            .unwrap();
        assert_eq!(active.id, profile.id);

        let cross_scope = set_active_style_profile(
            &conn,
            SetActiveStyleProfileInput {
                project_id: "11111111-1111-1111-1111-111111111111".to_string(),
                style_profile_id: profile.id.clone(),
            },
        );
        assert!(cross_scope.is_err());
        let still_active = get_active_style_profile(&conn, "missing-project")
            .unwrap()
            .unwrap();
        assert_eq!(still_active.id, profile.id);

        delete_style_profile(&conn, "", &profile.id).unwrap();
        assert!(list_style_profiles(&conn, None)
            .unwrap()
            .iter()
            .all(|item| item.id != profile.id));
    }
}
